pippy/PipelineStage.py

```diff
@@ -653,6 +671,93 @@ def forward_one_chunk(
 
         return output, send_reqs
 
+    def forward_one_chunk_ipipe(
+        self,
+        chunk: int,
+        args_split,
+        kwargs_split,
+        fwd_cache: Dict[int, Any],
+    ):
+        if self.rank == self.nstages - 1:
+            # Need improvement- how do we properly pass targets to forward_maybe_with_nosync?
+            targets = args_split[chunk][0]
+            print(f"[Rank{self.rank}] args_split {args_split}")
+        else:
+            targets = None
+
+        composite_args, composite_kwargs = self._recv_and_fill_inputs(
+            chunk,
+            args_split,
+            kwargs_split,
+        )
+
+        try:
+            if self.rank == self.nstages - 1:  # last stage
+                output = self.forward_maybe_with_nosync(
```

Comment:
Do you need a inner_depth == 0 check here?

Reply:
no we don't. inner_depth == 1 is default (no inner pipelining) and when inner_depth == 1, this function is not called. Instead, it calls forward_one_chunk function.